### PR TITLE
support container to container copy

### PR DESCRIPTION
**Pull request by vrothberg**
Implement container to container copy.  Previously data could only be
copied from/to the host.

Fixes: #7370
Co-authored-by: Mehul Arora <aroram18@mcmaster.ca>
Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
@containers/podman-maintainers PTAL

File: cmd/podman/containers/cp.go
Comment (rhatdan) [inline cmd/podman/containers/cp.go:135]: NIT: Why not just use `err` here?

File: cmd/podman/containers/cp.go
Comment (rhatdan) [inline cmd/podman/containers/cp.go:170]: This is strange syntax, why not break into two lines?

**Review by rhatdan — APPROVED** [review]

File: cmd/podman/containers/cp.go
Comment (rhatdan) [reply cmd/podman/containers/cp.go:135]: I see below you used the err, but it is very hard to follow.  Could you create a boolean variable with a name indicating what you intend.

**Comment by edsantiago** [comment]
It looks like the ten-minute problem is caused by #10701: there are a lot of `podman exec`s in these tests, and each one takes 25-30 seconds.

File: cmd/podman/containers/cp.go
Comment (TomSweeneyRedHat) [inline cmd/podman/containers/cp.go:223]: Does the file system ensure that we don't write over a container we don't own?

Comment (vrothberg) [reply cmd/podman/containers/cp.go:223]: Yes.  We can only use containers in our local storage.

File: cmd/podman/containers/cp.go
Comment (vrothberg) [reply cmd/podman/containers/cp.go:135]: Can I tackle this and the other comment in another PR?  Now it's consistent with the other two copy functions but I can spin up a follow-up PR trying to simplify things a bit more.

**Comment by vrothberg** [comment]
> It looks like the ten-minute problem is caused by #10701: there are a lot of `podman exec`s in these tests, and each one takes 25-30 seconds.

Yes, the tests are noticeably slow.  They used to  _much_ faster.  I have absolutely no idea what may be the cause, so I'll try bisecting.

**Comment by vrothberg** [comment]
> Yes, the tests are noticeably slow. They used to run _much_ faster. I have absolutely no idea what may be the cause, so I'll try bisecting.

I do not notice a difference between Podman v3.0.1-rhel and v3.2.3 on my machine.  Could it be conmon?

**Comment by vrothberg** [comment]
> > Yes, the tests are noticeably slow. They used to run _much_ faster. I have absolutely no idea what may be the cause, so I'll try bisecting.
> 
> I do not notice a difference between Podman v3.0.1-rhel and v3.2.3 on my machine. Could it be conmon?

Scratch that.  Found a reproducer and the difference is very noticeable.

**Comment by vrothberg** [comment]
I stop spamming here now.  Let's continue over in #10701.

I think we should wait for #10701 to be fixed before merging this PR.

**Comment by vrothberg** [comment]
With the commit from https://github.com/containers/podman/pull/11055, the execution time of the cp tests on my machine dropped from ~8 min down to 3min 40sec.

I will still look into reducing the number of tests a bit.

**Comment by vrothberg** [comment]
Rebased on top of the exec fixes.  I also took a shot at reducing the exec's in the cp tests and cut down the execution time by around 25 percent on my machine.

**Comment by edsantiago** [comment]
The exec changes cut test time on my laptop from 3:46 to 3:36 (4% improvement). How did you get 25%? Are you comparing to old-slow-broken-exec time?

**Comment by vrothberg** [comment]
> The exec changes cut test time on my laptop from 3:46 to 3:36 (4% improvement). How did you get 25%? Are you comparing to old-slow-broken-exec time?

Old & slow: > 8min
This PR without cutting down execs: 3min 40sec
This with with cutting down execs: 2min 50sec

**Comment by edsantiago** [comment]
LGTM!

**Comment by rhatdan** [comment]
/lgtm
Thanks @infiniteregrets and @vrothberg

**Pull request merged**
